File: app/src/main/java/com/example/articles/components/ApplicationComponent.java
package com.example.articles.components;


import com.example.articles.ApplicationContext;
import com.example.articles.domain.repositories.IRepoFactory;
import com.example.articles.presentation.configuration.IPresentationConfigurator;
import com.example.articles.presentation.configuration.PresentationConfigurator;
import com.example.articles.presentation.articles.list.view.MainActivity;

import javax.inject.Singleton;

import dagger.Component;

@Singleton // Constraints this component to one-per-application or unscoped bindings.
@Component(
        modules = { ApplicationModule.class }
)

public interface ApplicationComponent{

    void inject( ApplicationContext application );
    void inject( MainActivity mainActivity );
    void inject( PresentationConfigurator presentationConfigurator );

    IRepoFactory repoFactory();
    IPresentationConfigurator presentationConfigurator();
}

File: app/src/main/java/com/mlukov/articles/di/components/ApplicationComponent.kt
package com.mlukov.articles.di.components


import com.mlukov.articles.ArticlesApp
import com.mlukov.articles.di.modules.*

import javax.inject.Singleton

import dagger.BindsInstance
import dagger.Component

@Singleton // Constraints this component to one-per-application or unscoped bindings.
@Component(
        modules = arrayOf(ApplicationModule::class,
                ActivityModule::class,
                DeviceModule::class,
                NetworkModule::class,
                BuildModule::class)
)
interface ApplicationComponent {

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun application(app : ArticlesApp) : Builder
        fun network(networkModule : NetworkModule) : Builder
        fun device( deviceModule : DeviceModule) : Builder
        fun build(): ApplicationComponent
    }

    fun inject(application : ArticlesApp)
}
